docs/rfcs/0003-initial.md

```diff
@@ -0,0 +1,280 @@
+      the last query was of a type that could not affect any rows, the result of
+      this method may be a stale value or `undefined`.
+  - `.close[Sync](): void`
+    - close the connection.
+- `driver.Transaction` interface
+  - `.query[Sync](sql: string, arguments?: Array<Value>): AsyncIterable<Iterable<Value>>`
+    - Executes a query against the database in the context of this transaction,
+      returning the results as an `AsyncIterable` of `Iterable` rows of
+      `Value`s. These iterables will not be used after the associated
+      transaction has ended.
+  - `.commit[Sync](): void`
+    - Ends the transaction, with any committed and saved. The transaction object
+      will not be used again.
+  - `.rollback[Sync](): void`
+    - Ends the transaction, with any changes rolled back. The transaction object
+      will not be used again.
+  - `.startTransaction[Sync](): driver.Transaction`
+    - Starts a new child transaction within this transaction. The transaction
+      object will not be used again until the child transaction has ended.
+
+### Driver `Meta` Type Information
+
+A driver SHOULD declare an associated `Meta` type using the `driver.Meta<{...}>`
+type function. This is currently used only to specify the `Value` type used by
+the driver for bound and result column values. The `Meta` type SHOULD be
+provided as the (optional) first type argument to every interface that is
+implemented from `driver.sql`.
+
+```ts
+type Meta = driver.Meta<{
+  Value: bigint | number | null
+}>;
+
+export class Driver extends driver.Driver<Meta> { … }
+```
+
+If a driver does not define and use its own `Meta` type, a default `Meta` will
+be used which expects only the the JSON primitive types: `null`, `boolean`,
+`number` and `string`.
+
+## Database Consumer API (`x/database/sql/sql.ts`)
+
+Inspired by [`sql.go`](https://golang.org/src/database/sql/sql.go), this is
+primary interface, which most users will use. `Value` below represents the
+driver-defined `Value` type for bindings and results. The interface is entirely
+async for now, even if the underlying driver supports sync operations.
+
+- `sql.open(path, driver): Promise<sql.Database>`
+  - Creates a database handle/connector with the given driver and path. May
+    validate the arguments (path), but will not open a connection yet.
+- `sql.Database` class
+  - `.connect(): Promise<sql.Connection>`
+    - Opens a new open connection to the database. In the future a connection
+      pool will be maintained instead of always opening new connections.
+- `sql.Connection` class
+  - `.startTransaction(): Promise<sql.Transaction>`
+    - Starts a new transaction in the connection. If a top-level transaction is
+      already in progress on this connection, this will block until it is
+      finished.
+  - `.close(): Promise<void>`
+    - Closes the connection. If there is an active transaction, this will block
+      until it is finished.
```

Comment:
There should also be a method for executing a query without having to start a transaction, as this is not always wanted/needed.
```suggestion
- `sql.Connection` class
  - `.query(args?): AsyncGenerator<Iterator<Value>>`
    - Executes a query without starting a transaction
  - `.startTransaction(): Promise<sql.Transaction>`
    - Starts a new transaction in the connection. If a top-level transaction is
      already in progress on this connection, this will block until it is
      finished.
  - `.close(): Promise<void>`
    - Closes the connection. If there is an active transaction, this will block
      until it is finished.
```

Reply:
Ah, interesting, I wasn't sure whether or not that was a requirement.

I guess that some database drivers don't allow DDL statements inside of transactions, and that's a pretty important case! Do you happen to know of any others?

Reply:
Added a `.prepareStatement()` method to `Connection` for executing queries without a transaction.